launchable/commands/record/build.py

```diff
@@ -169,100 +212,127 @@ def build(ctx: click.core.Context, build_name: str, source: List[str], max_days:
+    def synthesize_workspaces() -> List[Workspace]:
+        # workspace by its name
+        ws = {}  # type: Dict[str,Workspace]
 
         commit_pattern = re.compile("[0-9A-Fa-f]{5,40}$")
-        for repo_name, hash in _commits:
+
+        branch_name_map = dict(normalize_key_value_types(branches))
+
+        for name, hash in normalize_key_value_types(commits):
             if not commit_pattern.match(hash):
                 click.echo(click.style(
-                    "{}'s commit hash `{}` is invalid.".format(repo_name, hash),
+                    "{}'s commit hash `{}` is invalid.".format(name, hash),
                     fg="yellow"),
                     err=True)
-                invalid = True
-            submodules.append((repo_name, "", hash))
-        if invalid:
-            sys.exit(1)
+                sys.exit(1)
 
-    # Note: currently becomes unique command args and submodules by the hash.
-    # But they can be conflict between repositories.
-    uniq_submodules = list({commit_hash: (name, repo_dist, commit_hash)
-                            for name, repo_dist, commit_hash, in sources + submodules}.values())
+            ws[name] = Workspace(name=name, commit_hash=hash, branch=branch_name_map.get(name))
 
-    if no_commit_collection and len(branches) != 0:
-        branch_name_map = dict(normalize_key_value_types(branches))
-        repo_names = [m[0] for m in uniq_submodules]
-        not_match_branch_repos = set(branch_name_map.keys()) - set(repo_names)
-        if len(not_match_branch_repos) > 0:
-            click.echo(
-                click.style(
-                    "--branch option is set for {} but was not set for the --commit option".format(not_match_branch_repos),
+        # make sure no invalid branch name options are given
+        for k in branch_name_map:
+            if not ws.get(k):
+                click.echo(click.style(
+                    "Invalid repository name {} in a --branch option. "
+                    "Perhaps you missed the corresponding --commit option?".format(k),
                     fg="yellow"),
-                err=True)
-
-    build_id = None
-    tracking_client = TrackingClient(Tracking.Command.RECORD_BUILD)
-    try:
-        commitHashes = [{
-            'repositoryName': name,
-            'commitHash': commit_hash,
-            'branchName': branch_name_map.get(name, "")
-        } for name, _, commit_hash in uniq_submodules]
-
-        if not (commitHashes[0]['repositoryName'] and commitHashes[0]['commitHash']):
-            sys.exit('Please specify --source as --source .')
-
-        payload = {
-            "buildNumber": build_name,
-            "commitHashes": commitHashes,
-        }
-
-        _links = capture_link(os.environ)
-        if len(links) != 0:
-            for link in normalize_key_value_types(links):
+                    err=True)
+                # TODO: is there any reason this is not an error? for now erring on caution
+                # sys.exit(1)
```

Comment:
Calling an attention to here.

The original code was only treating this as a warning and not an error, and there's even a test case for that, so I assume it was important to behave that way, but I can't think of why.

This seems to be pretty squarely fall on the side of incorrect integration that should be an error, not a warning-&-recovery.